src/radiation/radiation_system.hpp

```diff
@@ -1962,6 +1983,15 @@ void RadSystem<problem_t>::AddSourceTerms(array_t &consVar, arrayconst_t &radEne
 			consNew(i, j, k, x3RadFlux_index + numRadVars_ * g) = Frad_t1[2][g];
 		}
 	});
+
+	const int nf_coupling = *(num_failed_coupling.copyToHost());
+	const int nf_dust = *(num_failed_dust.copyToHost());
```

Comment:
I'm not sure whether `copyToHost` is asynchronous or not. Can you check this? If it's asynchronous, then there is a race condition here: we don't know whether the GPU will write to the memory location before the CPU reads it.

For performance reasons, it would be better to do this after all of the FABs have been processed, rather than after each FAB. Otherwise, the GPU will have to synchronize with the CPU after each FAB (in order to guarantee that the CPU is reading the up-to-date value of each variable).

Reply:
I tested this PR on gadi GPU and it indeed aborted as expected when I set max_Iter to a small number.

Reply:
> For performance reasons, it would be better to do this after all of the FABs have been processed, rather than after each FAB.

I'll see if I can do that. My idea is to set a counter in `QuokkaSimulation<problem_t>::subcycleRadiationAtLevel` and pass it to `operatorSplitSourceTerms` then `AddSourceTerms`.

Reply:
Ok, it's good to also check the source code: https://github.com/AMReX-Codes/amrex/blob/c061a14011d4d58755f330d15a20f7243863a90d/Src/Base/AMReX_GpuBuffer.H#L91.

There is still a potential performance issue. Maybe you could allocate the `GpuBuffer`s in the calling function, and then check the result there?

Reply:
Updated.

I'm not sure what you mean by "calling function". Now, `copyToHost` is called once per radiation cycling. This should be optimal enough. I also tested it on gadi GPU and the code aborted as expected when set to.